src/retention/memory.rs

```diff
@@ -0,0 +1,554 @@
+    pub(crate) async fn set_capacity(&self, capacity: usize) {
+        self.store.lock().await.set_capacity(capacity);
+    }
+}
+
+#[derive(Debug)]
+struct VolatileStore {
+    store: VecDeque<VolitileItem>,
+}
+
+impl VolatileStore {
+    fn new() -> Self {
+        Self {
+            store: VecDeque::new(),
+        }
+    }
+
+    #[cfg(test)]
+    fn with_capacity(capacity: usize) -> Self {
+        Self {
+            store: VecDeque::with_capacity(capacity),
+        }
+    }
+
+    fn push<T>(&mut self, id: Id, value: T)
+    where
+        T: TryInto<ItemValue, Error = VolitileItemError>,
+    {
+        if self.is_full() {
+            // remote the expired only if its full, it will be done while iterating
+            self.remove_expired();
+
+            // If still full, remove the oldest one
+            if self.is_full() {
+                self.store.pop_front();
+            }
+        }
+
+        let res = value.try_into();
+        debug_assert!(res.is_ok(), "BUG: value should have retention volatile");
+
+        let item = match res {
+            Ok(item) => item,
+            Err(err) => {
+                error!("{err}");
+
+                return;
+            }
+        };
+
+        self.store.push_back(VolitileItem::new(id, item));
+    }
+
+    fn mark_sent(&mut self, id: &Id, sent: bool) -> Option<bool> {
+        self.store
+            .iter_mut()
+            .find(|item| item.id == *id)
+            .map(|item| std::mem::replace(&mut item.sent, sent))
+    }
+
+    fn mark_received(&mut self, id: &Id) -> Option<ItemValue> {
+        let idx = self
+            .store
+            .iter()
+            .enumerate()
+            .find_map(|(idx, item)| (item.id == *id).then_some(idx))?;
+
+        self.store.remove(idx).map(|item| item.value)
+    }
+
+    fn pop_next(&mut self) -> Option<ItemValue> {
+        let now = SystemTime::now();
+
+        std::iter::from_fn(|| self.store.pop_front())
+            .find(|item| !item.is_expiered(now))
+            .map(|item| item.value)
+    }
+
+    fn remove_expired(&mut self) {
+        let now = SystemTime::now();
+
+        self.store.retain(|item| !item.is_expiered(now));
+    }
+
+    fn is_full(&mut self) -> bool {
+        self.store.len() == self.store.capacity()
+    }
+
+    fn set_capacity(&mut self, capacity: usize) {
+        let current = self.store.capacity();
+
+        if capacity < current {
+            let diff = current.saturating_sub(capacity);
+            // Remove first elements
+            self.store.drain(..diff);
+
+            self.store.shrink_to_fit();
+        } else {
+            // Number of elements that needed to be reserved
+            let additional = capacity.saturating_sub(self.store.len());
+
+            self.store.reserve_exact(additional);
+        }
+    }
+}
+
+impl Default for VolatileStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct VolitileItem {
```

Comment:
Why didn't you reuse `PublishInfo` struct from the module for the items stored in volatile memory? Maybe some lines of code can be eliminated.

Reply:
I don't immediately see where the PublishInfo could be used for the volatile store.

The VolatileItem has additional information for the store time and sent flag, which are missing from the PublishInfo. Furthermore, the PublishInfo requires the payload to be serialized befeore storing it, while for the VolatileItem we store the AstarteType or Object.